### PR TITLE
[Fix] UNO 44 - Match interaction defining correct answer

### PR DESCRIPTION
**Pull request by habemuscode**
**Related to**

https://oat-sa.atlassian.net/browse/UNO-44

**Client Description**

- Match interaction is added
- Selecting response
- Selecting map response
- Giving scores to some answer-possibilities
- However regretting this and rather wanting "Match correct"
- Selecting match correct

-> in response it is not possible to choose correct answers, need to go back to map response, select "defining correct answer", and then choose

When user in 'Match' interaction select 'Map response' as response processing then regrets it and sets 'Match correct' as response processing, the choices get grayed and cursor becomes 'forbidden' when hovered over the checkbox, however user is able to set correct response and save it. 
To make checkboxes available again user has to go back to 'Map response', select 'defining correct answer', choose 'Match correct' and then choose. 

**How to reproduce**

- Add Match interaction to the canvas
- Set 'Map response' as response processing
- Set 'Match correct' as response processing
- Define the correct response

**Experienced Behavior**

Choices got grayed and the mouse pointer turns into Forbidden cursor icon when hovered over the choice checkbox (but the checkbox can be selected).
Expected Behavior

When 'Match correct' is set after 'Map response' was chosen as response processing, the Response mode turns to the default state: checkboxes are not grayed and user can select any of them as response.

### CODE REVIEW

File: views/js/qtiCreator/widgets/interactions/matchInteraction/states/Map.js [hunk 1]
```diff
@@ -17,30 +17,35 @@ define([
 
         //finally, apply defined correct response and response mapping:
         responseWidget.setResponse(interaction, _.values(response.getCorrect()));
-        
-       //change the correct state 
-       _widget.on('metaChange', function(meta){
+
+        $('.match-interaction-area input[type="checkbox"]', _widget.$container)
```
Comment (adriloma) [inline views/js/qtiCreator/widgets/interactions/matchInteraction/states/Map.js:21]: As `$('.match-interaction-area input[type="checkbox"]', _widget.$container)` is a recurrent selector on your code I would use a constant instead.

`const interaction = $('.match-interaction-area input[type="checkbox"]', _widget.$container);`

File: views/js/qtiCreator/widgets/interactions/matchInteraction/states/Map.js [hunk 2]
```diff
@@ -17,30 +17,35 @@ define([
...
+            .prop('disabled', true)
+            .prop('checked', false)
+            .addClass('disabled');
+
+       //change the correct state
+        _widget.on('metaChange', function(meta){
```
Comment (adriloma) [inline views/js/qtiCreator/widgets/interactions/matchInteraction/states/Map.js:27]: Have you applied `prettier`?
`_widget.on('metaChange', function(meta) {`

**Review by adriloma — CHANGES_REQUESTED** [review]
Code review only.

In addition to inline comments, and considering this is a small file, I would suggest to use this PR to reduce the tech debt and upgrade it to TAO best practises (remove the use of `var`, replace anoynmous functions by arrow functions if possible etc)

- [ ] New code is covered by tests (if applicable)
- [ ] Tests are running successfully (old and new ones) on my local machine (if applicable)
- [x] New code is respecting code style rules
- [ ] New code is respecting best practices
- [ ] New code is not subject to concurrency issues (if applicable)
- [ ] Feature is working correctly on my local machine (if applicable)
- [x] Acceptance criteria are respected
- [x] Pull request title and description are meaningful
- [x] Pull request's target is not `master`
- [x] bumping versions are applied (if applicable)

File: views/js/qtiCreator/widgets/interactions/matchInteraction/states/Map.js (hunk 2 above)
Comment (habemuscode) [reply views/js/qtiCreator/widgets/interactions/matchInteraction/states/Map.js]: As I know and see in `package.json` we don't have prettier ( maybe people use globally? ). So I put all file with same look

File: views/js/qtiCreator/widgets/interactions/matchInteraction/states/Map.js (hunk 1 above)
Comment (habemuscode) [reply views/js/qtiCreator/widgets/interactions/matchInteraction/states/Map.js]: Done :)

File: views/js/qtiCreator/widgets/interactions/matchInteraction/states/Map.js [hunk 3]
```diff
@@ -4,44 +4,47 @@ define([
...
-
+        $('.match-interaction-area input[type="checkbox"]', this.widget.$container)
+            .removeProp('disabled')
+            .removeClass('disabled');
         responseWidget.destroy(this.widget);
-
     });
-
-    return MatchInteractionStateMap;
 });
```
Comment (zagovorichev) [inline views/js/qtiCreator/widgets/interactions/matchInteraction/states/Map.js:50]: Could you add new line, I don't like this icon :(

**Review by zagovorichev — CHANGES_REQUESTED** [review]
It works on my environment as expected.
Bump the version, please

File: views/js/qtiCreator/widgets/interactions/matchInteraction/states/Map.js (hunk 3 above)
Comment (habemuscode) [reply views/js/qtiCreator/widgets/interactions/matchInteraction/states/Map.js]: Done

File: views/js/qtiCreator/widgets/interactions/matchInteraction/states/Map.js [hunk 4]
```diff
@@ -4,44 +4,47 @@ define([
...
 
-    var MatchInteractionStateMap = stateFactory.create(Map, function(){
+    return stateFactory.create(Map, function() {
 
-        var _widget = this.widget,
+        const _widget = this.widget,
             interaction = _widget.element,
-            response = interaction.getResponseDeclaration();
+            response = interaction.getResponseDeclaration(),
+            matchInteractionArea = $('.match-interaction-area input[type="checkbox"]', _widget.$container);
```
Comment (jsconan) [inline views/js/qtiCreator/widgets/interactions/matchInteraction/states/Map.js:14]: Better to split by one declaration per variable:
```javascript
        const _widget = this.widget;
        const response = interaction.getResponseDeclaration();
        const matchInteractionArea = $('.match-interaction-area input[type="checkbox"]', _widget.$container);
```

And you can also take the chance to fix the code style issue with the name of the variable `_widget`. Names must not start with a `_`.

**Review by jsconan — COMMENTED** [review]
The issue is fixed. 
Just one remark regarding code style. Since you updated some lines to ES6, you should also take the chance to fix the code style on them.

**Review Checklist**
- [ ] New code is covered by tests (if applicable)
- [x] Tests are running successfully (old and new ones) on my local machine (if applicable)
- [ ] New code is respecting code style rules
- [x] New code is respecting best practices
- [x] New code is not subject to concurrency issues (if applicable)
- [x] Feature is working correctly on my local machine (if applicable)
- [x] Acceptance criteria are respected
- [x] Pull request title and description are meaningful
- [x] Pull request's target is not `master`

File: views/js/qtiCreator/widgets/interactions/matchInteraction/states/Map.js (hunk 4 above)
Comment (habemuscode) [reply views/js/qtiCreator/widgets/interactions/matchInteraction/states/Map.js]: Done, take a look if is better name :)

**Review by jsconan — APPROVED** [review]
**Review Checklist**
- [ ] New code is covered by tests (if applicable)
- [x] Tests are running successfully (old and new ones) on my local machine (if applicable)
- [x] New code is respecting code style rules
- [x] New code is respecting best practices
- [ ] New code is not subject to concurrency issues (if applicable)
- [x] Feature is working correctly on my local machine (if applicable)
- [x] Acceptance criteria are respected
- [x] Pull request title and description are meaningful
- [x] Pull request's target is not `master`

**Review by zagovorichev — APPROVED** [review]
- [ ] New code is covered by tests (if applicable)
- [x] Tests are running successfully (old and new ones) on my local machine (if applicable)
- [x] New code is respecting code style rules
- [ ] New code is respecting best practices
- [ ] New code is not subject to concurrency issues (if applicable)
- [ ] Feature is working correctly on my local machine (if applicable)
- [ ] Acceptance criteria are respected
- [ ] Pull request title and description are meaningful

**Pull request merged**